sound/soc/codecs/rt5682.c

```diff
@@ -2521,13 +2521,15 @@ static int rt5682_wclk_prepare(struct clk_hw *hw)
 	struct rt5682_priv *rt5682 =
 		container_of(hw, struct rt5682_priv,
 			     dai_clks_hw[RT5682_DAI_WCLK_IDX]);
-	struct snd_soc_component *component = rt5682->component;
-	struct snd_soc_dapm_context *dapm =
-			snd_soc_component_get_dapm(component);
+	struct snd_soc_component *component;
+	struct snd_soc_dapm_context *dapm;
 
 	if (!rt5682_clk_check(rt5682))
 		return -EINVAL;
 
+	component = rt5682->component;
+	dapm = snd_soc_component_get_dapm(component);
```

Comment:
do you actually need this specific change @shumingfan ? I don't really see why it's necessary?

Reply:
@plbossart Sorry, I forgot to explain more about this PR.
It will call the callback function when doing the devm_clk_hw_register function in the clk.c.
Like this 
https://github.com/thesofproject/linux/blob/e1e85cff35eb5c92ab618216c052bb2fb669f307/drivers/clk/clk.c#L3544
Due to moving DAI clock registration to i2c probe, the codec component doesn't ready to access.
To avoid the NULL pointer reference, the driver assigns the component variable after the rt5682_clk_check.

Reply:
sorry @shumingfan, I really don't get the flow. 

rt5682->component is initialized in rt5682_probe(), which is a component probe. So if you use this field in a check done when doing the i2c probe, it is not initialized yet.